cpp/serve/engine_actions/new_request_prefill.cc

```diff
@@ -117,83 +120,130 @@ class NewRequestPrefillActionObj : public EngineActionObj {
-          embeddings = input_data[i]->GetEmbedding(models_[model_id],
-                                                   /*dst=*/!single_input ? &embeddings : nullptr,
-                                                   /*offset=*/cum_prefill_length);
-          cum_prefill_length += input_data[i]->GetLength();
+        RECORD_EVENT(trace_recorder_, prefill_inputs[i].rsentry->request->id, "start embedding");
+        // Speculative models shift left the input tokens by 1 when base model has committed tokens.
+        // Note: for n > 1 cases Eagle doesn't work because parent entry doesn't shift input tokens.
+        int embed_offset =
+            prefill_inputs[i].rsentry->mstates[model_id]->committed_tokens.empty() ? 0 : 1;
+        for (int j = 0; j < static_cast<int>(input_data.size()); ++j) {
+          if (j == static_cast<int>(input_data.size()) - 1) {
+            std::vector<int32_t> tail_tokens;
+            TokenData tk_data = Downcast<TokenData>(input_data[j]);
+            for (int k = embed_offset; k < static_cast<int>(tk_data->token_ids.size()); ++k) {
+              tail_tokens.push_back(tk_data->token_ids[k]);
+            }
+            embeddings = models_[model_id]->TokenEmbed(
+                {tail_tokens.begin(), tail_tokens.end()},
+                /*dst=*/!single_input ? &model_workspaces_[model_id].embeddings : nullptr,
+                /*offset=*/cum_prefill_length);
+            cum_prefill_length += input_data[j]->GetLength();
+            cum_prefill_length -= embed_offset;
+          } else {
+            embeddings = input_data[i]->GetEmbedding(
+                models_[model_id],
+                /*dst=*/!single_input ? &model_workspaces_[model_id].embeddings : nullptr,
+                /*offset=*/cum_prefill_length);
+            cum_prefill_length += input_data[j]->GetLength();
+          }
+        }
+        if (embed_offset > 0) {
+          std::vector<int32_t> new_tokens = {prefill_inputs[i]
+                                                 .rsentry->mstates[model_id]
+                                                 ->committed_tokens.back()
+                                                 .sampled_token_id.first};
+          embeddings =
+              models_[model_id]->TokenEmbed({new_tokens.begin(), new_tokens.end()},
+                                            /*dst=*/&model_workspaces_[model_id].embeddings,
+                                            /*offset=*/cum_prefill_length);
+          cum_prefill_length += new_tokens.size();
         }
         RECORD_EVENT(trace_recorder_, rsentry->request->id, "finish embedding");
       }
 
       RECORD_EVENT(trace_recorder_, request_ids, "start prefill");
-      NDArray logits =
-          models_[model_id]->BatchPrefill(embeddings, request_internal_ids, prefill_lengths);
+      NDArray hidden_states = models_[model_id]->FuseEmbedHidden(
+          embeddings, hidden_states_for_input, /*batch_size*/ 1, /*seq_len*/ cum_prefill_length);
```

Comment:
I'm a bit confused. By using `hidden_states_for_input` here, is it undefined for the model 0, but defined for model 1?

Reply:
Yes. Only speculative small models need the previous hidden states. Base model is responsible to produce the hidden states.